src/librustc/traits/specialize/specialization_graph.rs

```diff
@@ -291,8 +335,52 @@ impl<'a, 'gcx, 'tcx> Graph {
 
     /// The parent of a given impl, which is the def id of the trait when the
     /// impl is a "specialization root".
-    pub fn parent(&self, child: DefId) -> DefId {
-        *self.parent.get(&child).unwrap()
+    pub fn parent(&self, child: DefId) -> Vec<DefId> {
+        self.parent.get(&child).unwrap().clone()
+    }
+
+    pub fn build_graph(&self) -> (DataStructuresGraph<DefId, String>, DefIdMap<NodeIndex>) {
+        let mut sg_graph: DataStructuresGraph<DefId, String> = DataStructuresGraph::new();
+        let mut nodes_idx = Default::default();
+        for (key, val) in self.parent.iter() {
+            let idx = self.node_idx(&mut sg_graph, &mut nodes_idx, *key);
+            for parent in val {
+                let pidx = self.node_idx(&mut sg_graph, &mut nodes_idx, *parent);
+                sg_graph.add_edge(idx, pidx, format!("fromt {:?} to {:?}", key, parent));
+            }
+        }
+
+        (sg_graph, nodes_idx)
+    }
+
+    /// Return true if impl1 and impl2 are allowed to overlap:
+    /// They have an intersection impl
+    pub fn check_overlap(&self,
+                         sg_graph: &DataStructuresGraph<DefId, String>,
+                         nodes_idx: &DefIdMap<NodeIndex>,
+                         impl1: DefId,
+                         impl2: DefId) -> bool {
```

Comment:
Nit: can we format this method the rustfmt way?

Reply:
I've run the rustfmt tool on the entire file. It seems that there were other methods not formatted.